server/lib/pbench/process_tb.py

```diff
@@ -0,0 +1,113 @@
+        and copies it to the remote server by creating a subprocess which
+        call Agent's 'pbench-results-push' command.
+
+            Args -
+                config -- PbenchServer config object
+                logger -- logger object to use when emitting log entries during
+                            operation
+
+        """
+        self.config = config
+        self.logger = logger
+        # receive_dir -- directory where tarballs are received from agent
+        self.receive_dir = self.get_receive_dir()
+
+    def get_receive_dir(self) -> Path:
+        receive_dir_prefix = self.config.get(
+            "pbench-server", "pbench-receive-dir-prefix"
+        )
+        if not receive_dir_prefix:
+            raise ValueError(
+                "Failed: No value for config option pbench-receive-dir-prefix in section pbench-server"
+            )
+
+        receive_dir = Path(f"{receive_dir_prefix}-002").resolve()
+        if not receive_dir.is_dir():
+            raise NotADirectoryError(
+                "Failed: {} does not exist, or is not a directory", receive_dir
+            )
+
+        return receive_dir
+
+    def results_push(self, controller: str, tb: Path, token: str):
+        """Runs Agent's `pbench-results-push` command with controller, tb
+        and token options
+
+        Args -
+            controller -- the name of the controller to be associated with the
+                          tar ball
+            tb -- path of the tar ball
+            token -- generated authorised token for Pbench user
+
+        Importance of this function is while running tests we get the
+        ability to mock this function and test it easily
+        """
+
+        res = subprocess.run(
+            ["pbench-results-push", controller, tb, token], capture_output=True
+        )
+        error = res.stderr
+        if res.returncode > 0:
+            raise RuntimeError(error)
+
+    def process_tb(self) -> Results:
+        """Searches for tarballs in the configured receive directory and
+        uploads them to the Pbench server"""
+
+        # Check for results that are ready for processing: version 002 agents
+        # upload the MD5 file as xxx.md5.check and they rename it to xxx.md5
+        # after they are done with MD5 checking so that's what we look for.
+        self.logger.info("{}", self.config.TS)
+        token = self.config.get("pbench-server", "put-token")
+        nstatus = ""
+        ntotal = ntbs = 0
+
+        for tbmd5 in sorted(self.receive_dir.glob("**/*.tar.xz.md5")):
+            ntotal += 1
+
+            # Extracts full pathname of tarball from the name of hashfile
+            # by trimming .md5 suffix
+            tb = Path(str(tbmd5)[0 : -len(".md5")])
+            tbdir = tb.parent
+            controller = tbdir.name
+
+            try:
+                self.results_push(controller, tb, token)
+            except Exception as e:
+                self.logger.error(
```

Comment:
We are already checking the actual error, it is output by `e` and is reported by the `results_push` function.

Reply:
True.

However, if the error is that tar ball already exists, that might indicate a more systemic problem: "why would this shim be pushing the same tar ball multiple times?"  Could it be the result of errors below?  Or if a user keeps pushing the same tar ball by mistake, should we be reporting it as an error?  Or should be report it separately as a duplicate which is already on the server?

Reply:
yes, I agree but how will we be able to decide that the tarball has already been sent? The only way I can see, is by looking at logs, if the tarball is failing to get removed only then it will be still present in the `results-receive-directory` and we can guess that by looking at `errors encountered` in status or logs.

Reply:
The logs are a quasi-output of a function.  As such, it might be appropriate for unit test code to inspect them, in order to ensure that the CUT is functioning correctly, but it should not be the case that production code depends on being able to read and interpret the logs.  We **should** be able to find or build a better mechanism than that!

Reply:
Take a look at `lib/pbench/server/api/resources/upload_api.py`, and you'll see the status codes it returns.  If a tar ball has already been uploaded, it returns a 200 (`HTTPStatus.OK`) and returns a 201 (`HTTPStatus.CREATED`) if it was not a duplicate.

If the API returns a 400 (`HTTPStatus.BAD_REQUEST`), then we should probably stop processing any more tar balls, because this client is not working with the server; similarly with `HTTPStatus.LENGTH_REQUIRED`.

If the API returns a `HTTPStatus.REQUEST_ENTITY_TOOL_LARGE`, then we need to quarantine the tar ball to the side to have it reported later.

If the API returns a `HTTPStatus.INSUFFICIENT_STORAGE`, then we should stop sending tar ball as well.

If the API returns a 5xx error, then there is likely something wrong with the pbench server, and the client should backoff and try again later.

Reply:
The problem is, the code here in `put-shim` isn't invoking that interface, not even indirectly, so it doesn't, as things stand, have any hope of getting ahold of those status returns, as far as I know.

This is not to say that we couldn't alter that reality, but the fact is, as currently constructed, `put-shim` invokes `pbench-results-push`, and `pbench-results-push` doesn't return those values.  In fact, I'm not even sure that _it_ has access to them.

So, this is going to lead to the question, which interface _should_ `put-shim` be invoking?  I think that that is a good and worthy question, but, in the spirit of being agile, I think it is one that we should defer to another PR.

Reply:
This remains a vector where the put shim can repeatedly send the same tar ball to the production server.  If the tar ball is large enough, it will be a drain on the server.

Deferring this to another PR is okay if we open an issue to track this work.

Reply:
#2450 I have created an issue for this to keep track.